### ISSUE TITLE
[SQL] runtime errors should contain source position information

### ISSUE THREAD

**Issue by mihaibudiu** [feldera/feldera#1177]
The compiler side should generate code for all operations that contains source position information, similar to the way code is generated for UDFs. Currently Calcite does not propagate source position information to the back-end, but I have submitted a patch which will enable it to do that: https://github.com/apache/calcite/pull/3506. 

This will entail changing the signature in the Rust code in all the sqllib functions to take an extra source position argument.

Other steps will involve passing information about the source program to the runtime, and designing a mechanism for actually reporting the errors.

**Comment by ryzhyk** [comment]
I don't think that Rust functions (both library functions and UDFs) should be aware of location information generate by the compiler.  A Rust function must return a `Result<>` type containing an error message or descriptor with information specific to this function. The runtime should enrich these errors with location information and propagate them to the user.

**Comment by mihaibudiu** [comment]
So let's have a design before we start working on this.

**Comment by abhizer** [comment]
@ryzhyk How would the run time receive the location information when the rust functions don't get the source position information?

**Comment by ryzhyk** [comment]
The Rust function call is generated by the SQL compiler, which knows how to map the error back to the source position, so it should be able to wrap the function call to add the location information to the error.

**Comment by mihaibudiu** [comment]
Can you type a mock of the API you would expect?

**Comment by ryzhyk** [comment]
There are two APIs really. First, there's the interface between the DBSP library and user-provided closures. Consider, for instance, the `join` operator:

```rust
    pub fn join<I2, F, V>(
        &self,
        other: &Stream<C, I2>,
        join_func: F,
    ) -> Stream<C, OrdZSet<V, I1::R>>
    where
        I2: IndexedZSet<Key = I1::Key, R = I1::R> + Send,
        F: Fn(&I1::Key, &I1::Val, &I2::Val) -> V + Clone + 'static,
        V: DBData,
```

It takes a closure of type `Fn(&I1::Key, &I1::Val, &I2::Val) -> V`. This signature doesn't allow the function to fail, so any runtime errors will manifest as panic's, which will cause the DBSP runtime to shutdown.  We will change this signature to smth like `Fn(&I1::Key, &I1::Val, &I2::Val) -> Result<V, Error>`, where the details of the `Error` type are TBD, but I think it will be something akin to `anyhow::Error`, which essentially allows arbitrary user-defined error types to be returned.  In practice, the implementation of this closure is generated by the SQL compiler, which will include location information in the error details.

The second API is the interface between the Rust code generated by the SQL compiler and Rust UDFs (including SQL library functions, which @abhizer is currently working on), e.g.,

```rust
pub fn abs_i64(left: i64) -> i64
```

This signature will also change to `pub fn abs_i64(left: i64) -> Result<i64, UdfError>`, where the `UdfError` type is also TBD and it also needs to be flexible enough to fit all possible classes of errors, but it shouldn't include location information, since UDFs do not know where in the SQL code they are being invoked.

The glue between the two APIs is compiler-generated Rust code, which knows the SQL program location where the failing UDF is invoked and can inject this information into the error before returning it to DBSP.

Does this make sense?

**Comment by mihaibudiu** [comment]
The problem is if you have an expression such as a+b+c.
The way you suggest writing code will need to compile this into:

```
let tmp0 = add(a, b);
if !tmp0.ok() {
    return Err::new(position0);
}
let tmp1 = add(tmp0.unwrap(), c);
if !tmp1.ok() {
    return Err::new(position1);
}
tmp1.unwrap()
```

With the previous proposal the generated code is much nicer:

```
let tmp0 = add(a, b, position0)?;
let tmp1 = add(tmp0, c, position1)?;
tmp1
```

**Comment by mihaibudiu** [comment]
In fact, it's worse than that, you have to match on each tmp to extract the error and pass it forward.

**Comment by ryzhyk** [comment]
Yes, `map_err` is the way to go.

**Comment by mihaibudiu** [comment]
In the other proposed variant you can even do `add(add(a, b)?, c)`

**Comment by mihaibudiu** [comment]
So this is the proposal: `add(add(a, b).map_err(|e| Err::new(e, pos0))?, c).map_err(|e| Err::new(e, pos1))`

**Comment by mihaibudiu** [comment]
these are not even udfs, they are are own library functions

**Comment by ryzhyk** [comment]
yes, but UDFs will have the same calling convention.

**Comment by ryzhyk** [comment]
> So this is the proposal: `add(add(a, b).map_err(|e| Err::new(e, pos0))?, c).map_err(|e| Err::new(e, pos1))`

yes

**Comment by mihaibudiu** [comment]
It's not difficult for the compiler to generate this kind of code.
We should start by designing the error class and the Debug for it.

**Comment by abhizer** [comment]
Actually, with @ryzhyk 's method, we can also probably do `add(add(a, b)?, c)`. We would need to make something like a static map like [phf](https://docs.rs/phf/0.11.2/phf/), and include something identifying the function call, and the source position in it. 
And then, implement `From<Error> for ErrorWithPos` using the static map.

**Comment by mihaibudiu** [comment]
What is the key used to index in the map?

**Comment by ryzhyk** [comment]
> and include something identifying the function call, and the source position in it.

where would this identifier appear in the generated code?

**Comment by mihaibudiu** [comment]
I guess we are both asking the same question.

**Comment by abhizer** [comment]
I was thinking of something like `stringify!("{fn_name}{params}")` but this fails for duplicate invocation.

**Comment by ryzhyk** [comment]
> We should start by designing the error class and the Debug for it.

One option is [anyhow::Error](https://docs.rs/anyhow/latest/anyhow/struct.Error.html).

Pros:
- allows library functions to return any existing error type as long as it implements `Debug` and `Display`.

Cons:
- doesn't allow enforcing any additional structure on the error type, e.g., we won't be able to have an error type enum.

**Comment by abhizer** [comment]
> So this is the proposal: `add(add(a, b).map_err(|e| Err::new(e, pos0))?, c).map_err(|e| Err::new(e, pos1))`

I think we should put the `map_err` inside a macro. 

Consider the error type as such: 
```rust
pub struct ErrWithSource {
    err: String,
    src: SourcePositionRange,
}
```

And a macro as such: 

```rust
macro_rules! c {
    ($fn: expr, $source_pos: expr) => {{
        let res = $fn;
        let src: SourcePositionRange = $source_pos;

        // Won't be necessary in actual code, as it will return ErrorType
        let res = Ok::<_, String>(res);

        res.map_err(|e: String| ErrWithSource { err: e, src })
    }};
}
```

And the generated code can be: 

```rust
        let stream2: Stream<_, OrdZSet<Tuple1<Option<i64>>, Weight>> = INT8_TBL.map(
            move |t: &Tuple2<Option<i64>, Option<i64>>| -> Tuple1<Option<i64>> {
                let pos0 = SourcePositionRange::default();
                let pos1 = SourcePositionRange::default();
                let pos2 = SourcePositionRange::default();

                // we don't create this extra closure right now, but adding this would help bubble up errors
                let finalout = move || -> Result<Tuple1<Option<i64>>, ErrWithSource> {
                    {
                        Ok(Tuple1::new(c!(
                            plus_i64N_i64N(
                                c!(
                                    minus_i64N_i64N(
                                        (*t).0,                                    
                                        c!(plus_i64N_i64N((*t).0, (*t).1), pos0)?,
                                    ),
                                    pos1
                                )?,
                                (*t).0,
                            ),
                            pos2
                        )?))
                    }
                };
                finalout().unwrap()
            },
        );
```

Instead of: 
```rust
        let stream2: Stream<_, OrdZSet<Tuple1<Option<i64>>, Weight>> = INT8_TBL.map(
            move |t: &Tuple2<Option<i64>, Option<i64>>| -> Tuple1<Option<i64>> {
                let pos0 = SourcePositionRange::default();
                let pos1 = SourcePositionRange::default();
                let pos2 = SourcePositionRange::default();

                let finalout = move || -> Result<Tuple1<Option<i64>>, ErrWithSource> {
                    {
                        Ok(Tuple1::new(
                            Ok(plus_i64N_i64N(
                                Ok(minus_i64N_i64N(
                                    (*t).0, 
                                    Ok(plus_i64N_i64N((*t).0, (*t).1))
                                        .map_err(|err| ErrWithSource { err, src: pos0 })?,
                                ))
                                .map_err(|err| ErrWithSource { err, src: pos1 })?,
                                (*t).0,
                            ))
                            .map_err(|err| ErrWithSource { err, src: pos2 })?,
                        ))
                    }
                };
                finalout().unwrap()
            },
        );
```

**Comment by mihaibudiu** [comment]
@gz @ryzhyk @lalithsuresh @blp please comment on this proposed API

**Comment by lalithsuresh** [comment]
I don't have much context for the end-to-end implications for this API. It looks orthogonal to how we will bubble up runtime errors. If so, I'll defer to the others on how you achieve it.

**Comment by gz** [comment]
> The problem is if you have an expression such as a+b+c. The way you suggest writing code will need to compile this into:
> 
> ```
> let tmp0 = add(a, b);
> if !tmp0.ok() {
>     return Err::new(position0);
> }
> let tmp1 = add(tmp0.unwrap(), c);
> if !tmp1.ok() {
>     return Err::new(position1);
> }
> tmp1.unwrap()
> ```
> 
> With the previous proposal the generated code is much nicer:
> 
> ```
> let tmp0 = add(a, b, position0)?;
> let tmp1 = add(tmp0, c, position1)?;
> tmp1
> ```

I don't think we should add this source code info manually or through the compiler (consider someone writes a >1k LOC UDF we don't want them to manually add all source code positions in the errors)
There are existing mechanisms and libraries that automatically include source position (including full backtrace) for where an error in the code happens:

- `std::backtrace::Backtrace` https://doc.rust-lang.org/std/backtrace/
- support exists for this in https://docs.rs/thiserror/latest/thiserror/ or https://docs.rs/anyhow/latest/anyhow/

**Comment by mihaibudiu** [comment]
We want the SQL source position, not (only) the Rust source position.

**Comment by gz** [comment]
Since Mihai informed me that this is for SQL source information... I'd check if there is a way to add annotations directly to DWARF by annotating the rust code. Then you would have the SQL-to-rust mapping encoded in the final binary and can parse it during backtrace. I'm don't know though if rust supports this.

**Comment by abhizer** [comment]
Here `String` is just a placeholder for the error type.

**Comment by ryzhyk** [comment]
@gz , DWARF would be neat, maybe not necessary for this specific issue, but for instance for debugging panics. But I'm also not sure how to do it.

**Comment by mihaibudiu** [comment]
Filed a fix https://github.com/apache/calcite/pull/3506 with Calcite, but even after that is merged more work will be needed. And no one has reviewed it for many months.
